pkg/cli: Refactor CLI to support streaming formatted results

richardartoul opened the pull request:
This is my first attempt to address #13362

The basic idea is that printQueryOutput accepts an iterator now, and can "stream" results to the formatters as they arrive. In practice, most of the formatters accept all the rows at once or have a "render" function that requires all rows to be present in the case of the pretty printer, but once we have iteration set up we can then work on making the formatters stream friendly.

This is my first pull request to this repository so I imagine you will have lots of feedback. I'm happy to take it all into account and refactor this pull request!

richardartoul committed:
pkg/cli: Refactor CLI to support streaming formatted results

Fixes #13362

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15277)
<!-- Reviewable:end -->

benesch commented:
Hi @richardartoul—someone else will be along shortly to give you a proper code review, but in the meantime, I just wanted to say thanks! This is an awesome improvement.

knz commented:
Excellent work @richardartoul! Thanks again for your contribution.
See below for my comments.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/cli/format_table.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiUmw-Q43GpvfoIulTy:-KiUmw-R0pYYCAeHNFrs:b-4vslb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d8ac9128a60f1af1a04b46e439cd8c33d07ba77/pkg/cli/format_table.go#L32)):*
> ```Go
> )
> 
> type rowStrIter interface {
> ```

Please document this interface and the two structs that implement it using comments.

---

*[pkg/cli/format_table.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiUn2wJM6_izLyFQh--:-KiUn2wJM6_izLyFQh-0:b56ye2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d8ac9128a60f1af1a04b46e439cd8c33d07ba77/pkg/cli/format_table.go#L43)):*
> ```Go
> 
> func (iter *rowSliceIter) Next() (row []string, err error) {
> 	if iter.index < len(iter.allRows) {
> ```

It is more customary in Go to have the special case in the conditional and the common case  out, e.g.

```go
if iter.index >= len(iter.allRows) {
    return nil, io.EOF
}
...
```

---

*[pkg/cli/format_table.go, line 127 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiUnQNcLUJ2Ox3spkCs:-KiUnQNcLUJ2Ox3spkCt:b-8b8jfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d8ac9128a60f1af1a04b46e439cd8c33d07ba77/pkg/cli/format_table.go#L127)):*
> ```Go
> 		fallthrough
> 	case tableDisplayCSV:
> 		aLLRowsSlice, err := allRows.ToSlice()
> ```

"aLL" -> "all"

---

*[pkg/cli/sql_util.go, line 316 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiUo3clyM5msKLheC1p:-KiUo3cm8HilRJkJsYc4:bu2v0ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d8ac9128a60f1af1a04b46e439cd8c33d07ba77/pkg/cli/sql_util.go#L316)):*
> ```Go
> 	for {
> 		cols := getColumnStrings(rows)
> 		if len(cols) == 0 {
> ```

The comment does not exactly explain what is happening here...

FYI, you can have a table with 0 columns but multiple rows:
```
root@:26257/> CREATE TABLE t.foo (x INT);
CREATE TABLE
root@:26257/> insert into t.foo values(42),(69),(100);
INSERT 3
root@:26257/> select count(*) from t.foo;
+----------+
| count(*) |
+----------+
|        3 |
+----------+
(1 row)
root@:26257/> alter table t.foo drop column x;
ALTER TABLE
root@:26257/> SELECT count(*) FROM t.foo;
+----------+
| count(*) |
+----------+
|        3 |
+----------+
(1 row)
```

If you then ask the database for the contents of this table, you actually get 3 rows from the server, all of which with 0 columns. 

The reason for your special case here is that *although the table contains many rows, we want to improve the user's experience by only showing a summary, defined as the query tag followed by the row count*. And then, sinc the count visible via `RowsAffected()` (used by getFormattedTag) merely reports the number of rows successfully consumed by the driver, the client code very much has to consume all these empty rows for otherwise the final count would be incorrect.

Now my explanation is long but that's to explain you what's going on. It would be great if you could determine a suitable short summary.
(and possibly capture this logic in a separate function!)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-:-KiUpmIaZBk59dQGaYx3:b287kf1)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 4 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/cli/node.go, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiVL5svuU-f-I-m5-p7:-KiVL5svuU-f-I-m5-p8:b-5b2t9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d8ac9128a60f1af1a04b46e439cd8c33d07ba77/pkg/cli/node.go#L74)):*
> ```Go
> 	}
> 
> 	// printQueryOutput will not throw an error when provided a rowSliceIter
> ```

why not check it anyway?

---

*[pkg/cli/node.go, line 144 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiVL863-kO40nNOHuGG:-KiVL863-kO40nNOHuGH:b-rqxvo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d8ac9128a60f1af1a04b46e439cd8c33d07ba77/pkg/cli/node.go#L144)):*
> ```Go
> 	}
> 
> 	// printQueryOutput will not throw an error when provided a rowSliceIter
> ```

ditto.

---

*[pkg/cli/sql.go, line 266 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiVLBjNDC3mOEks9q0n:-KiVLBjNDC3mOEks9q0o:bh741gy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d8ac9128a60f1af1a04b46e439cd8c33d07ba77/pkg/cli/sql.go#L266)):*
> ```Go
> func (c *cliState) handleSet(args []string, nextState, errState cliStateEnum) cliStateEnum {
> 	if len(args) == 0 {
> 		// printQueryOutput will not throw an error when provided a rowSliceIter
> ```

prefer you panic here if you're sure it won't.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277)*
<!-- Sent from Reviewable.io -->

richardartoul commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/cli/node.go, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiVL5svuU-f-I-m5-p7:-KiYaymldwbYUsXVnDTy:b-xgjker) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d8ac9128a60f1af1a04b46e439cd8c33d07ba77/pkg/cli/node.go#L74)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why not check it anyway?
</blockquote></details>

I was just being shy about touching too much code ;) Will fix.

---

*[pkg/cli/sql_util.go, line 316 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiUo3clyM5msKLheC1p:-KiYbFH4Mci9Aly1agLA:bfgtgy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d8ac9128a60f1af1a04b46e439cd8c33d07ba77/pkg/cli/sql_util.go#L316)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The comment does not exactly explain what is happening here...

FYI, you can have a table with 0 columns but multiple rows:
```
root@:26257/> CREATE TABLE t.foo (x INT);
CREATE TABLE
root@:26257/> insert into t.foo values(42),(69),(100);
INSERT 3
root@:26257/> select count(*) from t.foo;
+----------+
| count(*) |
+----------+
|        3 |
+----------+
(1 row)
root@:26257/> alter table t.foo drop column x;
ALTER TABLE
root@:26257/> SELECT count(*) FROM t.foo;
+----------+
| count(*) |
+----------+
|        3 |
+----------+
(1 row)
```

If you then ask the database for the contents of this table, you actually get 3 rows from the server, all of which with 0 columns. 

The reason for your special case here is that *although the table contains many rows, we want to improve the user's experience by only showing a summary, defined as the query tag followed by the row count*. And then, sinc the count visible via `RowsAffected()` (used by getFormattedTag) merely reports the number of rows successfully consumed by the driver, the client code very much has to consume all these empty rows for otherwise the final count would be incorrect.

Now my explanation is long but that's to explain you what's going on. It would be great if you could determine a suitable short summary.
(and possibly capture this logic in a separate function!)
</blockquote></details>

Thanks for the detailed explanation :) I'll fix it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277)*
<!-- Sent from Reviewable.io -->

richardartoul commented:
Update my pull request with the change ya'll asked for, and also fixed a "bug" I had introduced where the showMoreChars was hard-coded as true instead of being configurable in the iterator.

If ya'll have more feedback I'm happy to make more changes.

Also, any thoughts on what would be a good new task to tackle thats a "step up" from this? I looked through the issues that were tagged as "easy" / "helpwanted" but didn't see anything that looked promising or wasn't already claimed.

knz commented:
Thanks for the improvements. This looks good.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/cli/format_table.go, line 32 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiZZ7sgTXeXBCePffIY:-KiZZ7shLPZ4bJeLILCV:b-jrhg61) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4037b07f82dbbefe7bd1008237ad0e3c41e0d43/pkg/cli/format_table.go#L32)):*
> ```Go
> )
> 
> // rowStrIter is an iterator interface for the printQueryOutput function. Its
> ```

"It is used"

---

*[pkg/cli/format_table.go, line 40 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiZZA4tVAub51suCQzo:-KiZZA4tVAub51suCQzp:b-s6v6ma) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4037b07f82dbbefe7bd1008237ad0e3c41e0d43/pkg/cli/format_table.go#L40)):*
> ```Go
> }
> 
> // rowSliceIter is an implementation of the rowStrIter interface and its used
> ```

"and it is used"

---

*[pkg/cli/format_table.go, line 61 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiZZH1dCdIDu2HZ-KNZ:-KiZZH1dCdIDu2HZ-KN_:b-s6v6ma) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4037b07f82dbbefe7bd1008237ad0e3c41e0d43/pkg/cli/format_table.go#L61)):*
> ```Go
> }
> 
> // newRowSliceIter is an implementation of the rowStrIter interface and its
> ```

"and it is used"

---

*[pkg/cli/node.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiZZVXhdlKLja7KT8hS:-KiZZVXivwIiw-ySrXvw:bsjh3am) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4037b07f82dbbefe7bd1008237ad0e3c41e0d43/pkg/cli/node.go#L74)):*
> ```Go
> 	}
> 
> 	err = printQueryOutput(os.Stdout, lsNodesColumnHeaders, newRowSliceIter(rows), "", cliCtx.tableDisplayFormat)
> ```

`return printQueryOutput(...)`

---

*[pkg/cli/node.go, line 145 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiZZf3MJWXwFgFrHOmX:-KiZZf3MJWXwFgFrHOmY:bsjh3am) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4037b07f82dbbefe7bd1008237ad0e3c41e0d43/pkg/cli/node.go#L145)):*
> ```Go
> 	err = printQueryOutput(os.Stdout, nodesColumnHeaders, newRowSliceIter(nodeStatusesToRows(nodeStatuses)), "",
> 		cliCtx.tableDisplayFormat)
> 	return err
> ```

`return printQueryOutput(...)`

---

*[pkg/cli/sql_util.go, line 322 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiZZxzjh4O8oPEyBQMx:-KiZZxzjh4O8oPEyBQMy:b-s8m2ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4037b07f82dbbefe7bd1008237ad0e3c41e0d43/pkg/cli/sql_util.go#L322)):*
> ```Go
> 			// method returns the correct number of rows (it only reports the number
> 			// of rows that the driver consumes).
> 			err := consumeAllRows(rows)
> ```

```go
if err := consumeAllRows(rows); err != nil {
   return err
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-:-KiZ_5KyreH9TyiUFFy9:b-2qspdi)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 4 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/cli/format_table.go, line 183 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KiZ0WmSAGTNv3YVJvD2:-KiZ0WmSAGTNv3YVJvD3:bljd30f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4037b07f82dbbefe7bd1008237ad0e3c41e0d43/pkg/cli/format_table.go#L183)):*
> ```Go
> 		}
> 
> 		i := 0
> ```

can hou scope `i` to the loop? `for i := 0;;i++ {`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277)*
<!-- Sent from Reviewable.io -->

richardartoul commented:
Cool, updated.

tamird commented:
Reviewed 1 of 3 files at r3.
Review status: 2 of 4 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/cli/sql_util.go, line 306 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-Kiez-Ics6XJbeL6nRSl:-Kiez-IdOht5fEUp5-FL:b-wmyxl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e814de80ed84234c62cf7d7e971e49b61c11d9a/pkg/cli/sql_util.go#L306)):*
> ```Go
> func runQueryAndFormatResults(
> 	conn *sqlConn, w io.Writer, fn queryFunc, displayFormat tableDisplayFormat,
> ) (err error) {
> ```

why the named return?

---

*[pkg/cli/sql_util.go, line 327 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-Kiez4nsmScYo1_Kj5wz:-Kiez4nsmScYo1_Kj5x-:b5v62b6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e814de80ed84234c62cf7d7e971e49b61c11d9a/pkg/cli/sql_util.go#L327)):*
> ```Go
> 
> 			fmt.Fprintln(w, getFormattedTag(rows.Tag(), rows.Result()))
> 			// Can return early because we know there are no more results
> ```

period at the end of the comment please.

---

*[pkg/cli/sql_util.go, line 330 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-Kiez1pVrazF-Sph7w-u:-Kiez1pVrazF-Sph7w-v:brdnukk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e814de80ed84234c62cf7d7e971e49b61c11d9a/pkg/cli/sql_util.go#L330)):*
> ```Go
> 			return nil
> 		}
> 		err := printQueryOutput(w, cols, newRowIter(rows, true), "", displayFormat)
> ```

`if err := ...; err != nil {`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a few minor comments. Thanks for your contribution!

Please also trim the commit message to omit the intermediate work (i.e. it can just read:
> pkg/cli: Refactor CLI to support streaming formatted results
> 
> Fixes #13362.

---

Reviewed 1 of 3 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/cli/sql_util.go, line 327 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-KikBI1I0VR8W0HjqlK6:-KikBI1I0VR8W0HjqlK7:bf7gjlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9703a810a4527952f963222e8595698ba5aab70/pkg/cli/sql_util.go#L327)):*
> ```Go
> 
> 			fmt.Fprintln(w, getFormattedTag(rows.Tag(), rows.Result()))
> 			// Can return early because we know there are no more results.
> ```

you can avoid this early return, right? In other words, if you just fall through to `printQueryOutput`, it will do the right thing - is that correct? assuming you remove printing the tag just above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-:-KikC3J5v68esLcslHME:b-5gbgkj)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> too 

---

Reviewed 2 of 3 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-:-Kil0D7zLW6pCTvZLpk-:bb7agjf)*
<!-- Sent from Reviewable.io -->

knz commented:
@richardartoul we're letting you have a last look to check whether you're happy, the result is good to go on our side.

richardartoul commented:
I'm ready for a merge :)

tamird commented:
Will merge on green. Thank you, @richardartoul!

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15277#-:-KivoIiIt7oS2JYRsvFA:b-scxm1d)*
<!-- Sent from Reviewable.io -->

mjibson commented:
This broke master:

```
pkg/cli/cert.go:189: cannot use rows (type [][]string) as type rowStrIter in argument to printQueryOutput:
	[][]string does not implement rowStrIter (missing Next method)
make: *** [build] Error 2
```
